thinc/layers/premap_ids.pyx

```diff
@@ -0,0 +1,85 @@
+# cython: infer_types=True
+cimport cython
+import numpy as np
+from preshed.maps cimport PreshMap
+from preshed.maps import PreshMap
+from typing import Dict, Union, Optional, cast, Callable, Tuple, Mapping
+from ..types import Ints1d, Ints2d
+from ..config import registry
+from ..model import Model
+from ..util import to_numpy
+
+
+InT = Union[Ints1d, Ints2d]
+OutT = Ints2d
+
+
+cdef lookup(PreshMap mapping, long[:] keys, long default):
+    """
+    Faster dict.get(keys, default) for the case when 
+    the "dict" is a Dict[int, int] converted to PreshMap
+    and the "keys" is a numpy integer vector.
+    """
+    cdef int maxi = len(keys)
+    result = np.empty(maxi, dtype="int")
+    cdef long[:] result_view = result
+    for i in range(maxi):
+        v = mapping[keys[i]]
+        if v is None:
+            result_view[i] = default
+        else:
+            result_view[i] = v
+    return result
+
+
+@cython.binding(True)
+@registry.layers("premap_ids.v1")
+def premap_ids(
+    mapping_table: Mapping[int, int],
+    default: int = 0,
+    *,
+    column: Optional[int] = None
+):
+    """Remap integer inputs to integers a mapping table, usually as a
+    preprocess before embeddings."""
+    if isinstance(mapping_table, PreshMap):
```

Comment:
Is it possible for `mapping_table` to have the type `PreshMap`?

Reply:
hmm okay. What I say in the type declaration is that it should be any `Mapping[int, int]` because all we need is for the `mapping_table` to implement `items()` and for that yield `Tuple[int, int]`. So I was thinking that users could give us any map like that including `PreshMap`. But I did realize now `PreshMap` is actually not a `Mapping`.

This might be a stupid question, but should be make `PreshMap` a `Mapping`? If not then I can make `mapping_table` type into `Union[Mapping[int, int], PreshMap]`.

Reply:
You could consider making `PreshMap` a `Mapping`, but I think it would be pretty low priority and might be more work than you'd expect. (`PreshMap` is kind of fiddly and I don't think `__len__` is correct right now, so things could go wrong fast.)

Reply:
I think for now I'd just keep it as `Mapping`. (I'd actually never initialized a `PreshMap` in python before testing this just now.)